src/Redux_Todos/TodoList.js

```diff
@@ -0,0 +1,112 @@
+import React, { useState } from "react";
+import { connect } from "react-redux";
+import { addTodo, deleteTodo, toggleTodo } from "./redux/actions";
+import InputComponent from "./InputComponent";
+
+const TodoList = ({ todos, addTodo, deleteTodo, toggleTodo }) => {
+  const [inputValue, setInputValue] = useState("");
+  const [filter, setFilter] = useState("all");
+  const [error, setError] = useState("");
+
+  const handleInputChange = (e) => {
+    setInputValue(e.target.value);
+  };
+
+  const handleAddTodo = () => {
+    if (inputValue.trim() === "") {
+      setError("Please enter a valid todo.");
+      return;
+    }
+
+    addTodo(inputValue);
+    setInputValue("");
+    setError("");
+  };
+
+  const handleDeleteTodo = (id) => {
+    deleteTodo(id);
+  };
+
+  const handleToggleTodo = (id) => {
+    toggleTodo(id);
+  };
+
+  const handleFilter = (filterType) => {
+    setFilter(filterType);
+  };
+
+  const filteredTodos = todos.filter((todo) => {
+    if (filter === "active") {
+      return !todo.isCompleted;
+    } else if (filter === "completed") {
+      return todo.isCompleted;
+    }
+    return true;
+  });
```

Comment:
When `filter === "all"` you don't need to do unnecessary loop `[].filter()`

So, I would change code to something like this

```
const filtered = filter === "all" ? todos : todos.filter(({ isCompleted }) => filter === "completed" ? isCompleted : !isCompleted)
```
===========
If you move `filter` state to redux, then you can do filtering inside `mapState`, such approach will be more optimized

```
const mapStateToProps = (state) => {
  const filtered = (
    state.filter === "all"
      ? state.todos
      : state.todos.filter(({ isCompleted }) => state.filter === "completed" ? isCompleted : !isCompleted)
  );

  return {
    todos: filtered
  };
};
```

Or even inline
```
const mapStateToProps = (state) => ({
  todos: (
    state.filter === "all"
      ? state.todos
      : state.todos.filter(({ isCompleted }) => state.filter === "completed" ? isCompleted : !isCompleted)
  )
});
```

Also, you can use destructuring for state
```
const mapStateToProps = ({ filter, todos }) => ({
  todos: filter === "all" ? todos : todos.filter(({ isCompleted }) => filter === "completed" ? isCompleted : !isCompleted)
});
```